// ng-simple-carousel/esm2015/lib/with-dots/with-dots.component.js
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class NgSimpleCarouselWithDotsComponent {
    constructor() {
        // @Input() infinite = false;
        this.imgChange = new EventEmitter();
    }
    ngOnInit() {
        this.activeId = this.images[0].id;
    }
    onImgChange(id) {
        this.activeId = id;
        this.imgChange.emit(id);
    }
    onDotsClick(id) {
        this.controller.jumpTo(id);
    }
}
NgSimpleCarouselWithDotsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-simple-carousel-with-dots',
                template: "<!--\n Copyright 2020 Google LLC\n\n Licensed under the Apache License, Version 2.0 (the \"License\");\n you may not use this file except in compliance with the License.\n You may obtain a copy of the License at\n\n      http://www.apache.org/licenses/LICENSE-2.0\n\n Unless required by applicable law or agreed to in writing, software\n distributed under the License is distributed on an \"AS IS\" BASIS,\n WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n See the License for the specific language governing permissions and\n limitations under the License.\n-->\n\n<div class=\"wrap\">\n  <ng-simple-carousel [images]=\"images\" [controller]=\"controller\" (imgChange)=\"onImgChange($event)\"\n    [objectFit]=\"objectFit\">\n  </ng-simple-carousel>\n  <div class=\"dots\">\n    <ng-container *ngFor=\"let image of images\">\n      <div class=\"dot\" (click)=\"onDotsClick(image.id)\">\n        <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"white\" width=\"100%\" height=\"100%\">\n          <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n          <path *ngIf=\"image.id === activeId\" d=\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2z\" />\n          <path *ngIf=\"image.id !== activeId\"\n            d=\"M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z\" />\n        </svg>\n      </div>\n    </ng-container>\n  </div>\n</div>\n",
                styles: [".wrap{position:relative;width:100%;height:100%}.dots{position:absolute;bottom:5px;display:flex;width:100%;justify-content:center}.dot{width:1rem;height:1rem;cursor:pointer}"]
            },] }
];
NgSimpleCarouselWithDotsComponent.ctorParameters = () => [];
NgSimpleCarouselWithDotsComponent.propDecorators = {
    controller: [{ type: Input }],
    images: [{ type: Input }],
    objectFit: [{ type: Input }],
    imgChange: [{ type: Output }]
};
