// ng-simple-carousel/esm2015/lib/with-button/with-button.component.js
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class NgSimpleCarouselWithButtonComponent {
    constructor() {
        // @Input() infinite = false;
        this.imgChange = new EventEmitter();
        this.isFirstImg = true;
        this.isLastImg = false;
    }
    ngOnInit() {
    }
    onPrev() {
        this.controller.prev();
    }
    onNext() {
        this.controller.next();
    }
    onImgChange(id) {
        if (id === this.images[0].id) {
            this.isFirstImg = true;
            this.isLastImg = false;
        }
        else if (id === this.images[this.images.length - 1].id) {
            this.isFirstImg = false;
            this.isLastImg = true;
        }
        else {
            this.isFirstImg = false;
            this.isLastImg = false;
        }
        this.imgChange.emit(id);
    }
}
NgSimpleCarouselWithButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-simple-carousel-with-button',
                template: "<!--\n Copyright 2020 Google LLC\n\n Licensed under the Apache License, Version 2.0 (the \"License\");\n you may not use this file except in compliance with the License.\n You may obtain a copy of the License at\n\n      http://www.apache.org/licenses/LICENSE-2.0\n\n Unless required by applicable law or agreed to in writing, software\n distributed under the License is distributed on an \"AS IS\" BASIS,\n WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n See the License for the specific language governing permissions and\n limitations under the License.\n-->\n\n<div class=\"wrap\">\n  <ng-simple-carousel [images]=\"images\" [controller]=\"controller\" (imgChange)=\"onImgChange($event)\"\n    [objectFit]=\"objectFit\">\n  </ng-simple-carousel>\n  <button class=\"button left\" (click)=\"onPrev()\" *ngIf=\"!isFirstImg\">\n    <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"#3e3e3e\" width=\"30px\" height=\"30px\">\n      <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n      <path d=\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z\" stroke=\"#616161\" /></svg>\n  </button>\n  <button class=\"button right\" (click)=\"onNext()\" *ngIf=\"!isLastImg\">\n    <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"#3e3e3e\" width=\"30px\" height=\"30px\">\n      <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n      <path d=\"M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z\" stroke=\"#616161\" /></svg>\n  </button>\n</div>\n",
                styles: [".wrap{position:relative;width:100%;height:100%}.button{border:none;border-radius:50%;width:30px;height:30px;background:hsla(0,0%,100%,.4117647058823529);position:absolute;top:calc(50% - 15px);padding:0;outline:none;cursor:pointer}.button.left{left:5px}.button.right{right:5px}"]
            },] }
];
NgSimpleCarouselWithButtonComponent.ctorParameters = () => [];
NgSimpleCarouselWithButtonComponent.propDecorators = {
    controller: [{ type: Input }],
    images: [{ type: Input }],
    objectFit: [{ type: Input }],
    imgChange: [{ type: Output }]
};
