// ng-simple-carousel/esm2015/lib/carousel.controller.js
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Subject } from "rxjs";
export class CarouselController {
    constructor() {
        this.jumpSub = new Subject();
        this.jump$ = this.jumpSub.asObservable();
        this.actionSub = new Subject();
        this.action$ = this.actionSub.asObservable();
    }
    jumpTo(id) {
        this.jumpSub.next(id);
    }
    prev() {
        this.actionSub.next('prev');
    }
    next() {
        this.actionSub.next('next');
    }
}
