// ng-simple-carousel/esm2015/lib/carousel.component.js
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export class NgSimpleCarouselComponent {
    constructor() {
        this.objectFit = 'contain';
        // @Input() infinite = false;
        this.imgChange = new EventEmitter();
        this.activeIndex = 0;
        this.scroll$ = new BehaviorSubject(false);
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        this.controller.jump$.pipe(takeUntil(this.destroy$)).subscribe(id => this.jumpTo(id));
        this.controller.action$.pipe(takeUntil(this.destroy$)).subscribe(action => action === 'prev' ? this.onPrev() : this.onNext());
    }
    ngAfterViewInit() {
        this.initObserve();
        this.startObserve();
    }
    initObserve() {
        const rootEl = this.containerEl.nativeElement;
        const options = {
            root: rootEl,
            rootMargin: '0px -50%',
        };
        this.observer = new IntersectionObserver((entries, observer) => {
            entries.forEach(entry => {
                if (entry.isIntersecting) {
                    const activeId = entry.target.id;
                    const newInd = this.images.findIndex(img => img.id === activeId);
                    this.onImgChange(activeId, newInd);
                }
            });
        }, options);
    }
    startObserve() {
        const rootEl = this.containerEl.nativeElement;
        Array.from(rootEl.children).forEach(img => this.observer.observe(img));
    }
    stopObserve() {
        this.observer.disconnect();
    }
    // Observer will call onImgChange()
    jumpTo(id) {
        const newInd = this.images.findIndex(img => img.id === id);
        if (this.activeIndex === newInd) {
            return;
        }
        const rootEl = this.containerEl.nativeElement;
        rootEl.scrollTo({
            top: 0,
            left: rootEl.offsetWidth * newInd,
            behavior: 'smooth'
        });
    }
    onPrev() {
        let newId;
        if (this.activeIndex - 1 >= 0) {
            newId = this.images[this.activeIndex - 1].id;
        }
        // if (this.infinite && this.activeIndex - 1 < 0) { newId = this.images[this.images.length - 1].id; }
        if (newId) {
            this.jumpTo(newId);
        }
    }
    onNext() {
        let newId;
        if (this.activeIndex + 1 < this.images.length) {
            newId = this.images[this.activeIndex + 1].id;
        }
        // if (this.infinite && this.activeIndex + 1 > this.images.length - 1) { newId = this.images[0].id; }
        if (newId) {
            this.jumpTo(newId);
        }
    }
    onImgChange(id, index) {
        if (this.activeIndex === index) {
            return;
        }
        this.activeIndex = index;
        this.imgChange.emit(id);
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
        this.stopObserve();
    }
}
NgSimpleCarouselComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-simple-carousel',
                template: "<!--\n Copyright 2020 Google LLC\n\n Licensed under the Apache License, Version 2.0 (the \"License\");\n you may not use this file except in compliance with the License.\n You may obtain a copy of the License at\n\n      http://www.apache.org/licenses/LICENSE-2.0\n\n Unless required by applicable law or agreed to in writing, software\n distributed under the License is distributed on an \"AS IS\" BASIS,\n WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n See the License for the specific language governing permissions and\n limitations under the License.\n-->\n\n<div class=\"ng-simple-carousel\" #container>\n  <div class=\"img_wrap\" *ngFor=\"let img of images; let i = index\" [id]=\"img.id\">\n    <picture>\n      <source *ngFor=\"let source of img?.sources\" [srcset]=\"source.srcset\" [type]=\"source.type\" [sizes]=\"source.sizes\">\n      <img class=\"img\" [class.activeImg]=\"i === activeIndex\" [class.prev]=\"i === activeIndex - 1\"\n        [class.next]=\"i === activeIndex + 1\" [src]=\"img.src\" [style.object-fit]=\"objectFit\" [srcset]=\"img.srcset\"\n        [alt]=\"img.alt\" [sizes]=\"img.sizes\" #image>\n    </picture>\n  </div>\n</div>\n",
                styles: [".ng-simple-carousel{max-width:100%;width:100%;height:100%;overflow-x:scroll;overflow-y:hidden;scroll-snap-type:x mandatory;display:flex;align-items:center;flex-flow:row nowrap;-ms-overflow-style:none;scrollbar-width:none}.ng-simple-carousel::-webkit-scrollbar{display:none}.img_wrap{flex-shrink:0;padding:0 1px}.img,.img_wrap{width:100%;height:100%}.img{scroll-snap-align:start;-o-object-fit:contain;object-fit:contain}"]
            },] }
];
NgSimpleCarouselComponent.ctorParameters = () => [];
NgSimpleCarouselComponent.propDecorators = {
    containerEl: [{ type: ViewChild, args: ['container',] }],
    controller: [{ type: Input }],
    images: [{ type: Input }],
    objectFit: [{ type: Input }],
    imgChange: [{ type: Output }]
};
